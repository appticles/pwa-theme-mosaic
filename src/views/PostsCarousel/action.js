import config from 'react-global-configuration';

export const REQUEST_POSTS = 'REQUEST_POSTS';
export const RECEIVE_POSTS = 'RECEIVE_POSTS';

export const requestPosts = () => ({
  type: REQUEST_POSTS,
});

export const receivePosts = posts => ({
  type: RECEIVE_POSTS,
  posts,
});

export const fetchPosts = (params = {}) => (dispatch) => {
  dispatch(requestPosts());
  config.set(window.__INITIAL_CONFIG__, { freeze: false });

  let url;
  if (params && params.id) {
    url = config.get('export').posts + '/' + String(params.id);
  } else {
    url =
      config.get('export').posts +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receivePosts(json)))
    .catch(() => {
      dispatch(receivePosts([]));
    });
};
