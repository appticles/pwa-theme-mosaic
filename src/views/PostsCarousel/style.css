.posts-carousel-container {
  position: absolute;
  width: 100%;
  height: 100%; }
  .posts-carousel-container .pusher,
  .posts-carousel-container .slick-slider,
  .posts-carousel-container .slick-list,
  .posts-carousel-container .slick-track {
    height: 100%; }
    .posts-carousel-container .pusher .slick-slide,
    .posts-carousel-container .slick-slider .slick-slide,
    .posts-carousel-container .slick-list .slick-slide,
    .posts-carousel-container .slick-track .slick-slide {
      display: flex;
      flex-direction: column; }
  .posts-carousel-container * {
    border-color: #dddddd !important; }
