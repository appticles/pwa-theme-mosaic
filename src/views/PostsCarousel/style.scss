@import "../../utilities/_variables.scss";

.posts-carousel-container {
  * {
    border-color: $border-color !important;
  }
  .slick-slide.slick-active {
    height: 100vh;
  }
}
