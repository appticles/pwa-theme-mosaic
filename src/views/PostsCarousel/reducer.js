import { combineReducers } from 'redux';
import Immutable from 'seamless-immutable';
import _ from 'lodash';
import PropTypes from 'prop-types';

import { REQUEST_POSTS, RECEIVE_POSTS } from './action';

export const postPropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  title: PropTypes.shape({
    rendered: PropTypes.string.isRequired,
  }).isRequired,
  author: PropTypes.number.isRequired,
  categories: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
  slug: PropTypes.string.isRequired,
  content: PropTypes.shape({
    rendered: PropTypes.string.isRequired,
    protected: PropTypes.bool.isRequired,
  }).isRequired,
  excerpt: PropTypes.shape({
    rendered: PropTypes.string.isRequired,
    protected: PropTypes.bool.isRequired,
  }).isRequired,
  date: PropTypes.string.isRequired, //TO DO : date validation => class Date
  // TO DO : proptype for image src (_embedded params)
});

export const INITIAL_STATE = Immutable({
  items: [],
  isFetching: 0,
  loadMore: true,
});

const items = (state = INITIAL_STATE.items, action) => {
  switch (action.type) {
    case REQUEST_POSTS:
      return state;
    case RECEIVE_POSTS:
      if (Array.isArray(action.posts)) {
        return _.unionBy(state, action.posts, 'id');
      }
      return _.unionBy(state, [action.posts], 'id');
    default:
      return state;
  }
};

const loadMore = (state = INITIAL_STATE.loadMore, action) => {
  switch (action.type) {
    case REQUEST_POSTS:
      return true;
    case RECEIVE_POSTS:
      if (Array.isArray(action.posts)) {
        return !(action.posts.length < 10);
      }
      return state;
    default:
      return state;
  }
};

const isFetching = (state = INITIAL_STATE.isFetching, action) => {
  switch (action.type) {
    case REQUEST_POSTS:
      return state + 1;
    case RECEIVE_POSTS:
      return state - 1;
    default:
      return state;
  }
};

export const getPosts = state => state.items;
export const getPostsByCategory = (state, categoryId) =>
  state.items.filter(post => post.categories.indexOf(Number(categoryId)) !== -1);

export const getLoadMorePosts = state => state.loadMore;
export const getPostsFetching = state => state.isFetching;

/**
 * Check the password protected status for a post.
 * Returns true if the post does not exist in the state.
 *
 * @param {Object} state
 * @param {Number} postId
 */
export const getPostProtectedStatus = (state, postId) => {
  if (!_.isEmpty(state.items)) {
    const post = _.find(state.items, postItem => Number(postItem.id) === Number(postId));

    return !_.isNil(post) ? post.content.protected : true;
  }

  return true;
};

export default combineReducers({
  items,
  isFetching,
  loadMore,
});
