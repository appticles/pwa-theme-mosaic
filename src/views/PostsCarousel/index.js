import React, { Component } from 'react';
import Slider from 'react-slick';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Loader, Header, Icon } from 'semantic-ui-react';

import { fetchPosts } from './action';
import { getPosts, postPropType, getPostsFetching } from './reducer';

import PostsList from './components/PostsList';
import Footer from '../../components/Footer/index';
import './style.css';

class PostsCarousel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      itemsPerCard: 2,
      pageNumber: 1,

      // if load more is enabled for loading more items
      loadMore: false,
    };

    this.loadMore = this.loadMore.bind(this);
  }

  componentWillMount() {
    this.readPosts(this.props.match.params.categoryId);
  }

  componentWillReceiveProps(nextProps) {
    // If the category has changed, get the posts for the new category
    if (this.props.match.params.categoryId !== nextProps.match.params.categoryId) {
      this.readPosts(nextProps.match.params.categoryId);
      return;
    }

    // If we receive more items, load more should be enabled for the next request
    if (nextProps.posts.length > this.props.posts.length) {
      this.setState({ loadMore: true });
    }
  }

  /**
   * Calculate the no of pages.
   */
  getNoPages() {
    const noPosts = this.props.posts.length;

    // Add +1 if the items are not exactly split over cards (last card is incomplete).
    return Math.round(noPosts / this.state.itemsPerCard) + (noPosts % this.state.itemsPerCard === 0 ? 0 : 1);
  }

  /**
   * Make request to load posts.
   * @param {Number} categoryId
   */
  readPosts(categoryId) {
    const { dispatch } = this.props;
    dispatch(
      fetchPosts({
        categories: categoryId,
        page: this.state.pageNumber,
        status: 'publish',
        per_page: this.state.itemsPerCard * 5,
      }),
    );

    this.setState({ pageNumber: this.state.pageNumber + 1, loadMore: false });
  }

  /**
   * Load more items when the carousel is swiped.
   * @param {Number} index = The index of the card.
   */
  loadMore(index) {
    if (this.state.loadMore === false) {
      return;
    }

    const noPages = this.getNoPages();

    if (index === noPages - 1) {
      this.readPosts(this.props.match.params.categoryId);
    }
  }

  /**
   * Split the posts on cards.
   * @param {Number} chunkSize
   */
  createPostsList(chunkSize) {
    const postsList = [];
    let i;

    for (i = 0; i < this.props.posts.length; i += chunkSize) {
      postsList.push(this.props.posts.slice(i, i + chunkSize));
    }
    return postsList;
  }

  render() {
    const listedPosts = this.createPostsList(2);
    const settings = {
      infinite: false,
      speed: 500,
      slidesToShow: 1,
      slidesToScroll: 1,
      afterChange: this.loadMore,
    };
    console.log('posts props', this.props, listedPosts.length);
    return (
      <div className="posts-carousel-container">
        {this.props.loading === 1 ? <Loader active /> : null}
        <Slider {...settings}>
          {this.props.loading === 0 && listedPosts.length === 0 ? (
            <div key={Math.random()}>
              <Header as="h3" icon textAlign="center" className="not-found">
                <Icon name="folder open" circular />
                <Header.Content>This category does not have any posts.</Header.Content>
              </Header>
              <Footer />
            </div>
          ) : (
            listedPosts.map(postsList => (
              <div key={Math.random()}>
                <PostsList postsList={postsList} category={this.props.match} />
                <Footer />
              </div>
            ))
          )}
        </Slider>
      </div>
    );
  }
}

PostsCarousel.propTypes = {
  dispatch: PropTypes.func.isRequired,
  posts: PropTypes.arrayOf(postPropType).isRequired,
  loading: PropTypes.number.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      categoryId: PropTypes.string.isRequired,
      categorySlug: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};
const mapStateToProps = state => ({
  posts: getPosts(state.posts),
  loading: getPostsFetching(state.posts),
});
function mapDispatchToProps(dispatch) {
  return Object.assign({ dispatch }, bindActionCreators({ fetchPosts }, dispatch));
}
export default connect(mapStateToProps, mapDispatchToProps)(PostsCarousel);
