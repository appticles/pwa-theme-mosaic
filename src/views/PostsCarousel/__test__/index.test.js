import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

import mockStore from '../../../__mocks__/redux-mock-store';
import PostsCarousel from '../index';

jest.mock('../components/PostsList/index', () => 'Posts list mock');
jest.mock('../../../components/Footer/index', () => 'Footer mock');

jest.mock('react-slick', () => require.requireActual('../../../__mocks__/react-slick').default);
jest.mock('react-global-configuration', () => require.requireActual('../../../__mocks__/react-global-configuration').default);

describe('Posts carousel', () => {
  it('renders posts container', () => {
    const store = mockStore({
      posts: {
        items: [
          {
            id: 53,
            title: {
              rendered: 'Post 1',
            },
            author: 56,
            categories: [32, 54, 6],
            slug: 'post-1',
            content: {
              rendered: 'Post mock content',
              protected: false,
            },
            excerpt: {
              rendered: 'Post excerpt',
              protected: false,
            },
            date: 'some mock date',
            _embedded: {
              'wp:term': [[
                { id: 32, name: 'Cat 1', slug: 'cat-1' },
                { id: 54, name: 'Some category', slug: 'some-category-slug-2' },
                { id: 6, name: 'Some category', slug: 'some-category-slug-32' },
              ]],
              author: [
                {
                  id: 56,
                  name: 'Some author',
                },
              ],
            },
          },
        ],
        isFetching: 0,
        loadMore: false,
      },
      categories: {
        items: [
          {
            id: 32,
            name: 'Cat 1',
            slug: 'cat-1',
          },
          {
            id: 43,
            name: 'Cat 2',
            slug: 'cat-2',
          },
        ],
        isFetching: 0,
      },
      translations: {
        items: {
          TEXTS: {},
        },
        isFetching: 0,
      },
    });

    const match = {
      params: {
        categoryId: '32',
        categorySlug: 'cat-1',
      },
    };

    expect(renderer.create(<PostsCarousel match={match} store={store} fetchPosts={jest.fn()} />)).toMatchSnapshot();
  });

  it('renders loader', () => {
    const store = mockStore({
      posts: {
        items: [],
        isFetching: 1,
        loadMore: false,
      },
      categories: {
        items: [],
        isFetching: 1,
      },
      translations: {
        items: {},
        isFetching: 1,
      },
    });

    const match = {
      params: {
        categoryId: '35',
        categorySlug: 'some-category-slug',
      },
    };
    expect(renderer.create(<PostsCarousel match={match} store={store} fetchCategories={jest.fn()} />)).toMatchSnapshot();
  });

  it('calculates the number of pages and cards with no post', () => {
    const match = {
      params: {
        categoryId: '35',
        categorySlug: 'some-category-slug',
      },
    };

    const componentInstance = renderer.create(
      <PostsCarousel.WrappedComponent
        dispatch={jest.fn()}
        posts={[]}
        loading={0}
        loadMore={false}
        match={match}
        category={undefined}
      />).getInstance();

    expect(componentInstance.getPageNumber()).toBe(0);
    expect(componentInstance.getNoCards()).toBe(0);
    expect(componentInstance.createPostsList(3)).toEqual([]);
  });

  it('calculates the number of pages and cards for single card', () => {
    const match = {
      params: {
        categoryId: '32',
        categorySlug: 'cat-1',
      },
    };

    const category = { id: 32, name: 'Cat 1', slug: 'cat-1' };

    const posts = [];

    for (let i = 1; i <= 2; i += 1) {
      posts.push(
        {
          id: i,
          title: {
            rendered: 'Post ' + String(i),
          },
          content: {
            rendered: 'Post content ' + String(i),
            protected: false,
          },
          excerpt: {
            rendered: 'Post excerpt',
            protected: false,
          },
          date: '2018-07-05 00:05:00',
          author: 3,
          categories: [32],
          slug: 'post-' + String(i),
          _embedded: {
            'wp:term': [[{ ...category }]],
            author: [{ id: 3, name: 'Some author' }],
          },
        });
    }

    const componentInstance = renderer.create(
      <PostsCarousel.WrappedComponent
        dispatch={jest.fn()}
        posts={posts}
        loading={0}
        loadMore={false}
        match={match}
        category={undefined}
      />).getInstance();

    expect(componentInstance.getPageNumber()).toBe(0);
    expect(componentInstance.getNoCards()).toBe(1);
    expect(componentInstance.createPostsList(3)).toEqual([[...posts]]);
  });

  it('calculates the number of pages and cards for multiple cards', () => {
    const match = {
      params: {
        categoryId: '32',
        categorySlug: 'cat-1',
      },
    };

    const category = { id: 32, name: 'Cat 1', slug: 'cat-1' };

    const posts = [];

    for (let i = 1; i <= 9; i += 1) {
      posts.push(
        {
          id: i,
          title: {
            rendered: 'Post ' + String(i),
          },
          content: {
            rendered: 'Post content ' + String(i),
            protected: false,
          },
          excerpt: {
            rendered: 'Post excerpt',
            protected: false,
          },
          date: '2018-07-05 00:05:00',
          author: 3,
          categories: [32],
          slug: 'post-' + String(i),
          _embedded: {
            'wp:term': [[{ ...category }]],
            author: [{ id: 3, name: 'Some author' }],
          },
        });
    }

    const componentInstance = renderer.create(
      <PostsCarousel.WrappedComponent
        dispatch={jest.fn()}
        posts={posts}
        loading={0}
        loadMore={false}
        match={match}
        category={undefined}
      />).getInstance();

    expect(componentInstance.getPageNumber()).toBe(1);
    expect(componentInstance.getNoCards()).toBe(5);
    expect(componentInstance.createPostsList(2)).toEqual(
      [
        [posts[0], posts[1]],
        [posts[2], posts[3]],
        [posts[4], posts[5]],
        [posts[6], posts[7]],
        [posts[8]],
      ],
    );
  });

  it('load more posts', () => {
    const match = {
      params: {
        categoryId: '32',
        categorySlug: 'cat-1',
      },
    };

    const category = { id: 32, name: 'Cat 1', slug: 'cat-1' };

    const posts = [];

    for (let i = 1; i <= 3; i += 1) {
      posts.push(
        {
          id: i,
          title: {
            rendered: 'Post ' + String(i),
          },
          content: {
            rendered: 'Post content ' + String(i),
            protected: false,
          },
          excerpt: {
            rendered: 'Post excerpt',
            protected: false,
          },
          date: '2018-07-05 00:05:00',
          author: 3,
          categories: [32],
          slug: 'post-' + String(i),
          _embedded: {
            'wp:term': [[{ ...category }]],
            author: [{ id: 3, name: 'Some author' }],
          },
        });
    }

    const dispatchSpy = sinon.spy(jest.fn());

    const componentInstance = renderer.create(
      <PostsCarousel.WrappedComponent
        dispatch={dispatchSpy}
        posts={posts}
        loading={0}
        loadMore
        match={match}
        category={category}
      />).getInstance();

    componentInstance.loadMore(1); // this call should dispatch to be called twice

    // dispatch is also called three times by the constructor
    expect(dispatchSpy.callCount).toBe(5);
  });
});
