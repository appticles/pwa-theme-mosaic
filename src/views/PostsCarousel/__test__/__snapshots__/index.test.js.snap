// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts carousel renders loader 1`] = `
<div
  className="posts-carousel-container"
>
  <div
    className="ui active loader"
  />
</div>
`;

exports[`Posts carousel renders posts 1`] = `
<div
  className="posts-carousel-container"
/>
`;
